zio-uring/src/main/scala/Ring.scala

```diff
@@ -66,26 +76,35 @@ class Ring(native: Native, ringFd: Long, completionsChunkSize: Int) {
     while (run) {
       val reqId   = resultBuffer.getLong()
       val retCode = resultBuffer.getInt()
-      val _   = resultBuffer.getInt()
+      val _       = resultBuffer.getInt()
       pendingReqs.remove(reqId) match {
-        case Callback.Read(buf, cb)    => cb(Chunk.fromByteBuffer(buf))
-        case c @ Callback.Write(cb)    => cb(c.readWritten(resultBuffer))
-        case c @ Callback.OpenFile(cb) => cb(retCode)
+        case Callback.Read(buf, cb) if retCode >= 0 => cb(Right(Chunk.fromByteBuffer(buf)))
+        case Callback.Read(buf, cb) if retCode < 0  => cb(Left(retCode))
+        case c @ Callback.Write(cb)                 => cb(retCode)
+        case c @ Callback.OpenFile(cb)              => cb(retCode)
         // Need a better way to know when we've read everything....
-        case null                      => run = false //sys.error(s"Oops: nonexistent request $reqId completed")
+        case null                                   => run = false //sys.error(s"Oops: nonexistent request $reqId completed")
       }
     }
   }
 
+  def cancel(requestId: Long): Unit = {
+    val reqId = requestIds.getAndIncrement()
+    pendingReqs.remove(requestId) match {
+      case null => ()
+      case c    => native.cancel(ringFd, reqId, requestId)
```

Comment:
Not sure what are the guarantees around cancellation on io_uring but I guess a canceled req could still show up on the completions queue, which would then throw you off in `peek` and `await` as the `reqId` would already be removed from the `pendingReqs` map.